MatchServer/MMatchRuleGunGame.cpp

```diff
@@ -0,0 +1,121 @@
+#include "stdafx.h"
+#include "MMatchRuleGunGame.h"
+#include "MBlobArray.h"
+#include <algorithm>
+#include <random>
+
+MMatchRuleGunGame::MMatchRuleGunGame(MMatchStage* pStage) : MMatchRule(pStage)
+{
+}
+
+MMatchRuleGunGame::~MMatchRuleGunGame()
+{
+}
+
+void MMatchRuleGunGame::OnBegin()
+{
+	m_MatchSet = MGetGunGame()->GetGGSet(rand() % MGetGunGame()->GetSetCount());
+	//To shuffle or not to shuffle? that is the question.
+	std::random_shuffle(m_MatchSet.begin(), m_MatchSet.end());
+}
+
+void MMatchRuleGunGame::OnRoundBegin()
+{
+
+}
+
+void MMatchRuleGunGame::OnEnd()
+{
+	m_MatchSet.clear();
+}
+
+void MMatchRuleGunGame::OnRoundEnd()
+{
+}
+
+bool MMatchRuleGunGame::RoundCount()
+{
+	if (++m_nRoundCount < 1) return true;
+	return false;
+}
+
+
+void MMatchRuleGunGame::OnRoundTimeOut()
+{
+	SetRoundArg(MMATCH_ROUNDRESULT_DRAW);
+}
+
+void MMatchRuleGunGame::OnGameKill(const MUID& uidAttacker, const MUID& uidVictim)
+{
+	MMatchObject* pAttacker = MGetMatchServer()->GetObject(uidAttacker);
+	MMatchObject* pVictim = MGetMatchServer()->GetObject(uidVictim);
+	if (!GetStage())
+		return;
+	if (!pAttacker || !pVictim)
+		return;
+
+	if (pAttacker == pVictim)
+		return;
+
+	SendNewSet(pAttacker->GetUID(), pAttacker->GetKillCount());
+}
+
+bool MMatchRuleGunGame::CheckKillCount(MMatchObject* pOutObj)
+{
+	MMatchStage* pStage = GetStage();
+	for (auto* pObj : pStage->GetObjectList())
+	{
+		if (pObj->GetEnterBattle() == false) continue;
+
+		if (pObj->GetKillCount() >= (unsigned int)pStage->GetStageSetting()->GetRoundMax())
+		{
+			pOutObj = pObj;
+			return true;
+		}
+	}
+	return false;
+}
+
+bool MMatchRuleGunGame::OnCheckRoundFinish()
+{
+	MMatchObject* pObj = nullptr;
+
+	if (CheckKillCount(pObj))
+	{
+		return true;
+	}
+	return false;
+}
+
+void MMatchRuleGunGame::SendNewSet(const MUID& uidPlayer,int KillCount)
+{
+	//TODO: fix this possible bug??
+	GGSet pSet = m_MatchSet.at(KillCount);
+	//if (pSet.WeaponSet[0] == 0 && pSet.WeaponSet[0]  == nullptr)
```

Comment:
Please do not leave commented code. If you think it is useful when debugging use #ifdef _DEBUG otherwise remove it completely.